growth-program/src/rewards/notes_history.rs

```diff
@@ -0,0 +1,89 @@
+use crate::project::ProjectId;
+
+use super::week_timekeeping::Week;
+
+multiversx_sc::imports!();
+multiversx_sc::derive_imports!();
+
+pub const MAX_NOTE_LENGTH: usize = 10;
+pub const MAX_NOTE_HISTORY: usize = 20;
+
+pub type NoteIndex = usize;
+pub type NoteData<M> = ManagedBuffer<M>;
+
+#[derive(TypeAbi, TopEncode, TopDecode, NestedEncode, NestedDecode)]
+pub struct Note<M: ManagedTypeApi> {
+    pub note_data: NoteData<M>,
+    pub week: Week,
+}
+
+#[multiversx_sc::module]
+pub trait NotesHistoryModule: crate::project::ProjectsModule {
+    #[view(getNotesHistory)]
+    fn get_notes_history(
+        &self,
+        project_id: ProjectId,
+        user: ManagedAddress,
+    ) -> MultiValueEncoded<Note<Self::Api>> {
+        self.require_valid_project_id(project_id);
+
+        let mut notes = MultiValueEncoded::new();
+        let user_id = self.user_ids().get_id(&user);
+        if user_id == NULL_ID {
+            return notes;
+        }
+
+        let first_note_index = self.first_note_index(project_id, user_id).get();
+        let current_note_index = self.current_note_index(project_id, user_id).get();
+        for i in (first_note_index..current_note_index).rev() {
+            let note = self.note(project_id, user_id, i).get();
+            notes.push(note);
+        }
+
+        notes
+    }
+
+    fn insert_note(&self, project_id: ProjectId, user_id: AddressId, note: &Note<Self::Api>) {
+        require!(note.note_data.len() <= MAX_NOTE_LENGTH, "Length too long");
+
+        let index_mapper = self.current_note_index(project_id, user_id);
+        let current_index = index_mapper.get();
+        self.note(project_id, user_id, current_index).set(note);
+        index_mapper.set(current_index + 1);
+
+        if current_index < MAX_NOTE_HISTORY {
+            return;
+        }
+
+        self.first_note_index(project_id, user_id)
+            .update(|first_note_index| {
+                self.note(project_id, user_id, *first_note_index).clear();
+                *first_note_index += 1;
+            });
+    }
+
+    #[storage_mapper("currentNoteIndex")]
+    fn current_note_index(
+        &self,
+        project_id: ProjectId,
+        user_id: AddressId,
+    ) -> SingleValueMapper<NoteIndex>;
+
+    #[storage_mapper("firstNoteIndex")]
+    fn first_note_index(
+        &self,
+        project_id: ProjectId,
+        user_id: AddressId,
+    ) -> SingleValueMapper<NoteIndex>;
+
+    #[storage_mapper("note")]
+    fn note(
```

Comment:
Wouldn't be more efficient to have a vecmapper for notes storage? With project_id and user_id as keys. So you can remove the first_note_index storage as well. So when you reach the cap, you just pop the first element and add the new one.
I think this would simplify the logic.

Reply:
VecMapper unfortunately doesn't have such features. You can't pop the first element, you can just remove it, which still maintains the original length.

Reply:
Yes, you're right, overlooked this. What about `SingleValueMapper<ManagedVec>`, which would work with what I've suggested, by popping the first element. You would also improve the `getNotesHistory` view function, as you would only read the storage once, and then just return the vector. Or create a MultiValueEncoded, but without reading the storage for each index. Or we could use the newly improved SetMapper, which now has new features like first, next or iter_from.
Not saying the current version is not good, but I think we would simplify the logic a bit if we got rid of the `note_index` from the key.